### PR TITLE
[stats] Refactor SqlIntegration and stats package to use sums

### PR DESCRIPTION
**Pull request by lukesonnet**
This PR refactors the `ExperimentMetricQueryResponse` to contain sums, sums of squares, sums of products, and other simple summary statistics in order to move more of the statistics logic into the python statistics library.

## Changes

It is intended in this version to have no impact on any experiment results, although there are a few very minor differences covered in the `testing` section

* Edits files in `back-end/src/integrations` to return sums, sums of squares, and sums of products instead of mean and stddev for MetricValue and ExperimentMetric queries. This allows us to delete a bunch of sql that was used to build covariances, variances, and contain logic about what kind of variance to estimate based on function parameters (e.g. those of [getVariationVariance](https://github.com/growthbook/growthbook/blob/854b30003b2b9f89d5e76e03c18ffb8b5ed840e1/packages/back-end/src/integrations/SqlIntegration.ts#L1120-L1151)).
* Modifies the type `ExperimentMetricQueryResponse` to handle these new fields, including more data about the statistic *and* metric types involved
* Modifies `getMetricValue()`, used on the Metrics pages, to also work with sums and sums of squares, but to process them in to mean and standard deviations within `processMetricValueQueryResponse` for now. This eventually will probably be handled by gbstats as we build out functionality on the metrics page.
* Refactors `gbstats.py` to be much simpler; we now no longer need helpers like `correctStdDev` (which actually are broken for non-sample mean statistics) as we can simply sum our sums, sums of squares, and sums of products when combining rows of data. Furthermore, we updated the `Statistic` models in `shared/models.py` to contain the variance logic mapped to the quantities that are needed to compute it, all under the appropriate type of statistic that gets built using some new factory methods (e.g. `variation_statistic_from_metric_row`)
* Accordingly, I bump the gbstats version to 0.4.0 and updated gen_notebook.py.

### Experiment results

I used the recently constructed query runner to test 6 engines across 84 different experiment-metric configurations (see this [Notion](https://www.notion.so/growthbook/Testing-Stats-Integrations-4e549c1591de444ea0f19473d3257f7d#24e7d771aece432f9aeeaad9e8aabfa7) for what was covered).

* **integration refactor returns same information**: The sums and sums of squares returned by this branch (when manually used to compute mean and stddev) perfectly match the mean and standard deviation from the main branch (down to 1e-13), except for in three cases:
  * Snowflake, where some remaining rounding error appears to cause differences of 5e-10 or less in means and therefore related standard deviations
  * Some standard deviations where the current branch is *wrong* because it thinks variance s undefined because it tries to apply `var(value)` on groups with only one user when actually there are some users with 0s that just aren't in the data. So this refactor fixes this edge case where there's only one user with a non-zero value and the standard deviation was wrongly returned as undefined.
  * There are some unexplained differences in `stddev` in mysql, but all differences are < 1e-10.
  
The above results can be seen in this notebook: https://colab.research.google.com/drive/1BOYPi0UdrgYFnVFqn9dBx-AraVusExm9?usp=sharing
  
  
  
* **gbstats returns the same experiment statistics**: Even though the inputs to gbstats are the same, gbstats itself is changing because now it needs to compute standard deviations for means, proportions, ratios, etc. So I updated some unit tests, but I also wanted to actually pass the query results through the new version of gbstats (and compare them to passing the old query results through the old version of gbstats). Here, again, results are promising.
  * Except for snowflake and one other edge case below, which has some cascading errors from the above rounding error, all results for variation means, standard deviations, percent chance of beating control, risk, p-values, uplift means, and uplift standard deviations are < 1e-8. Most errors that are between 1e-8 and 1e-13 are confined to ratio metrics in presto (this difference remains unexplained but seems negligible.
  *  The other edge case is caused by a bug in the prod implementation `correctStdDev`, which used to be used with `reduce_dimensionality` to combine aggregates across dimensions. This method provides slightly incorrect answers when applied to ratio metrics or to binomial metrics, as it only applies to sample mean standard deviations. So the refactor fixes this, but causes a discrepancy between main and prod.
  
The above results can be seen in this notebook: https://colab.research.google.com/drive/1GI9M5qoqRwE34QhAtTjS-BwThadvvm1I?usp=sharing, which compares results from [old gbstats and old integration](https://colab.research.google.com/drive/10xod6SXkMLe5DBBgSQU7V3yzXraw2iu-?usp=share_link) with [new gbstats and new integration](https://colab.research.google.com/drive/120VXPpIwVh9VvtcS91z0TKbIl3vRoViP?usp=share_link).

Together, the above two tests show that this integration change, when it comes to experiment results, produces no meaningful differences in results, and when there are any differences, they are the result of either fixing bugs or (I believe for some of the small, 1e-10 and below differences) an increase in precision since we do more complicated computations as late as possible in the analysis.

### Metric results

We don't have automated testing for the metrics page, but here are some screenshots showing binomial and count metrics working for postgres.

Revenue metric (main)
Screen Shot 2023-01-30 at 4 18 07 PM
Revenue metric (dev)
Screen Shot 2023-01-30 at 4 17 47 PM

Binomial metric (main)
Screen Shot 2023-01-30 at 4 18 45 PM

Binomial metric (dev)
Screen Shot 2023-01-30 at 4 19 06 PM

TODO:
- [X] Document that ratio metrics won't work for now with `getManualSnapshotData`, an edge case (I only added a comment in code; this is an edge use case for people manually entering data)
- [X] Ensure old NB still runs (decided to just let people use `gbstats` v 0.3.0 if they have old notebooks)
- [X] Rerun `gen_notebook()` test
- [x] Test in MixPanel and GA (jeremy tested in MixPanel, did not test GA thoroughly)

### CODE REVIEW

**Review by jdorn — CHANGES_REQUESTED** [review]

File: packages/back-end/src/integrations/Mixpanel.ts [hunk 1]
```diff
@@ -487,14 +493,14 @@ export default class Mixpanel implements SourceIntegrationInterface {
                 date: string;
                 count: number;
                 sum: number | null;
+                sum_squares: number | null;
```
[the remark this answers is not in the document]
Comment (lukesonnet) [reply packages/back-end/src/integrations/Mixpanel.ts:496]: Thank you. That whole `reduce` statement is a bit unclear to me. Perhaps we can discuss it briefly.

File: packages/back-end/src/integrations/Mixpanel.ts [hunk 2]
```diff
@@ -312,12 +312,16 @@ export default class Mixpanel implements SourceIntegrationInterface {
...
+          ];
           for(let i=1; i<row.value.length; i++) {
             ret.metrics.push({
               id: metricIds[i-1],
+              metric_type: metricTypes[i-1],
               count: row.value[i].count,
-              mean: row.value[i].avg,
-              stddev: row.value[i].stddev,
+              sum: row.value[i].sum,
+              sum_squares: row.value[i].sum_squares,
```
[the remark this answers is not in the document]
Comment (lukesonnet) [reply packages/back-end/src/integrations/Mixpanel.ts:324]: Yes, thank you. I'm correct that `row.value[i]` should be the output from `mixpanel.reducer.numeric_summary('value.m${i}')`, right?

File: packages/back-end/src/integrations/SqlIntegration.ts [hunk 3]
```diff
@@ -381,8 +375,8 @@ export default abstract class SqlIntegration
         , __overall as (
           SELECT
             COUNT(*) as count,
-            ${this.avg("coalesce(value,0)")} as mean,
-            ${this.stddev("coalesce(value,0)")} as stddev
+            COALESCE(SUM(value), 0) as main_sum,
+            COALESCE(SUM(POWER(value, 2)), 0) as main_sum_squares
```
[the remark this answers is not in the document]
Comment (lukesonnet) [reply packages/back-end/src/integrations/SqlIntegration.ts:379]: I'm not sure, but I think there's no reason not to have `COALESCE(X, 0)` throughout this section so I'm going to add it back in just to be safe.

File: packages/back-end/src/integrations/SqlIntegration.ts [hunk 4]
```diff
@@ -449,9 +443,18 @@ export default abstract class SqlIntegration
...
         users: parseInt(row.users) || 0,
-        count: parseFloat(row.count) || 0,
-        mean: parseFloat(row.mean) || 0,
-        stddev: parseFloat(row.stddev) || 0,
+        count: parseInt(row.count) || 0,
+        statistic_type: row.statistic_type ?? "",
+        main_metric_type: row.main_metric_type ?? "",
+        main_sum: parseFloat(row.main_sum) || 0,
+        main_sum_squares: parseFloat(row.main_sum_squares) || 0,
+        ...(row.denominator_metric_type && {
```
Comment (lukesonnet) [inline packages/back-end/src/integrations/SqlIntegration.ts:451]: @jdorn this is where I added the filter to remove unnecessary fields from returning in the UI if the metric is only a mean metric and not a ratio metric. See this loom for the updated, desired behavior: https://www.loom.com/share/da580ba3b0c543878f5e155c70e1183a

**Review by jdorn — DISMISSED** [review]

**Review by jdorn — APPROVED** [review]

**Review by jdorn — APPROVED** [review]

**Pull request merged**